src/lock.rs

```diff
@@ -0,0 +1,85 @@
+use std::fs::{self, write};
+use std::path::{Path, PathBuf};
+
+use serde::{Deserialize, Serialize};
+
+use crate::{files, util::Fallible};
+
+pub fn code_changed<P: AsRef<Path>>(lock_file: P, code_dir: P) -> Fallible<bool> {
+    if code_dir.as_ref().exists()
+        && code_dir.as_ref().is_dir()
+        && lock_file.as_ref().exists()
+        && lock_file.as_ref().is_file()
+    {
+        let mut code_hasher = Hasher::default();
+        code_hasher.consume_all(code_dir)?;
+        let code_hash = code_hasher.compute();
+
+        let lock = Lock::read(lock_file)?;
+
+        Ok(lock.code_hash != code_hash)
+    } else {
+        Ok(false)
+    }
+}
+
+/// Content for Yarner.lock files
+#[derive(Serialize, Deserialize)]
+pub struct Lock {
+    pub source_hash: String,
+    pub code_hash: String,
+}
+
+impl Lock {
+    pub fn read<P: AsRef<Path>>(path: P) -> Fallible<Self> {
+        let buf = files::read_file_string(path.as_ref())?;
+        let val = toml::from_str::<Self>(&buf)?;
+
+        Ok(val)
+    }
+
+    pub fn write<P: AsRef<Path>>(&self, path: P) -> Fallible {
+        let str = toml::to_string(self)?;
+        write(path, str)?;
+
+        Ok(())
+    }
+}
+
+pub struct Hasher {
```

Comment:
Does this and its inherent methods need to be `pub`?

Reply:
same as above